src/rust/catnap/win/overlapped.rs

```diff
@@ -0,0 +1,774 @@
+    yielder_handle: Option<YielderHandle>,
+    /// Set by the I/O processor to indicate overlapped result.
+    completion_key: usize,
+    /// A callback to free all resources associated with the completion for abandoned waits.
+    free: unsafe fn(*mut OVERLAPPED) -> (),
+    /// Ensure the data stays pinned since the OVERLAPPED must be pinned.
+    _marker: PhantomPinned,
+}
+
+/// The set of data which must live as long as the I/O operation, including any optional state from the caller.
+#[repr(C)]
+struct StatefulOverlappedCompletion<S> {
+    /// Inner data required by the completion processor; must be first to "downcast" to OVERLAPPED.
+    inner: OverlappedCompletion,
+    /// Caller state, encased in a cell for interior mutability.
+    state: UnsafeCell<S>,
+}
+
+/// This struct encapsulates the behavior of Windows I/O completion ports. This implementation exposes a single
+/// threaded interface for invoking and processing the results of overlapped I/O via idiomatic start-cancel-finish
+/// operations. While this type implements `Send`, multiple threads should avoid calling `process_events`, as there may
+/// be overhead or OS limitations on the number of such threads. This type cannot be shared among threads.
+pub struct IoCompletionPort {
+    /// The OS handle to the completion port.
+    iocp: HANDLE,
+
+    /// Marker to prevent this type from implementing `Sync`.
+    _marker: PhantomData<Cell<()>>,
+}
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+impl OverlappedResult {
+    /// Create an OverlappedResult from the completed OVERLAPPED structure and the completion key returned from the
+    /// I/O completion port.
+    pub fn new(overlapped: &OVERLAPPED, completion_key: usize) -> OverlappedResult {
+        Self {
+            completion_key,
+            result: NTSTATUS(overlapped.Internal as i32),
+            bytes_transferred: overlapped.InternalHigh as u32,
+        }
+    }
+
+    /// Convert the OverlappedResult into a Result<...> indicating success or failure of the referent operation.
+    pub fn ok(&self) -> Result<(), Fail> {
+        let win32_error: WIN32_ERROR = translate_ntstatus(self.result);
+        if win32_error.is_ok() {
+            Ok(())
+        } else {
+            Err(win32_error.into())
+        }
+    }
+}
+
+impl IoCompletionPort {
+    /// Create a new I/O completion port.
+    pub fn new() -> Result<IoCompletionPort, Fail> {
+        let iocp: HANDLE = match unsafe { CreateIoCompletionPort(INVALID_HANDLE_VALUE, None, 0, 1) } {
+            Ok(handle) => handle,
+            Err(err) => return Err(err.into()),
+        };
+
+        // Verified by windows crate.
+        assert!(!iocp.is_invalid());
```

Comment:
This will panic on failure path, is that ok? Or should we evaluate using debug_assert! here?